Bump `bitcoin` v0.32

DanGould opened the pull request:
At the [Rust Bitcoin Summit](https://github.com/rust-bitcoin/rust-bitcoin/discussions/3112) we decided the ecosystem will stay on rust-bitcoin 0.32 since BDK 1.0 uses it. So I'm trying to update payjoin and that requires upstream `bip21` to update.

DanGould committed:
Upgrade bitcoin to 0.32

InvalidNetworkError no longer reports address.network() since
that method was removed.

DanGould committed:
Release version 0.5.0

DanGould committed:
Bump MSRV to `1.56.1`

`rust-bitcoin` increased the MSRV to 1.56.1

tcharding commented on src/lib.rs:
```diff
@@ -140,10 +140,7 @@ impl<'a, T> Uri<'a, bitcoin::address::NetworkUnchecked, T> {
-                required: network,
-                found: *self.address.network(),
-            })
+            Err(InvalidNetworkError { required: network })
```
Perhaps we can get rid of this whole impl block now that we have `assume_checked` and `require_network` in `rust-bitcoin` now.

DanGould commented:
This recent push fixes Tobin's problem.

I spent half an hour trying to make this DRY but I'm not sure that you can since Uri construction moves `self` and `self.address.<CHECK>` also owns self. Maybe I'm just being silly but I'm a bit worn out.

```rs
    pub fn require_network(self, network: bitcoin::Network) -> Result<Uri<'a, bitcoin::address::NetworkChecked, T>, bitcoin::address::ParseError> {
        Ok(Uri {
            address: self.address.require_network(network)?,
            amount: self.amount,
            label: self.label,
            message: self.message,
            extras: self.extras,
        })
    }

    /// Marks URI validated without checks.
    pub fn assume_checked(self) -> Uri<'a, bitcoin::address::NetworkChecked, T> {
        Uri {
            address: self.address.assume_checked(),
            amount: self.amount,
            label: self.label,
            message: self.message,
            extras: self.extras,
        }
    }
```

Kixunil commented on src/lib.rs:
```diff
@@ -136,15 +136,14 @@ impl<'a, NetVal: NetworkValidation, T> Uri<'a, NetVal, T> {
+            label: self.label,
+            message: self.message,
+            extras: self.extras,
+        })
```
FYI I'm not sure why exactly I made my own type here but now it looks better since the resulting error is convertible to `UriError`.

DanGould replied:
it might make sense to actually return https://docs.rs/bitcoin/latest/bitcoin/address/struct.NetworkValidationError.html instead of the general ParseError

Kixunil replied:
Oh, yes, we should do that but also impl `From<NetworkValidationError> for UriError`.

Kixunil reviewed (commented):
Looks fine but I'd prefer to change the error type.

DanGould committed:
Handle require_network validation with de::Error

require_network always returns `ParseError` but conversion
to UriError to de::Error allows require_network and deserialization
to be handled in one function with one error variant.

tcharding reviewed (approved):
ACK 48190182894f23bc3f3996360468e59e2ca8e67d

Kixunil commented:
Oh, we got into a bit of a mess. First of all, I think the API should be consistent with that of `Address`.

So when I got suggestion by @DanGould that we should return `NetworkValidationError` I was thinking "of course, that's what the `bitcoin` API is actually returning, I remember this!" which was wrong and the reason I misremembered it is that while I did remember a PR that changed it, [I also remembered I objected to it. What I didn't remember is that I didn't object to it *strongly*](https://github.com/rust-bitcoin/rust-bitcoin/pull/2610#discussion_r1545192752) and I missed the fact that the PR got in before the release.

I'm really sorry about the noise and I'm quite unhappy how the API turned out but I guess it ended up being least bad - favoring a more common scenario over allowing niche code to be theoretically perfect. We can make up for it with good tests.

So I ask the changes I requested before to be reverted and instead return `UriError`.

DanGould commented:
> So I ask the changes I requested before to be reverted and instead return UriError.

Do you mean return `ParseError` ?

DanGould commented:
I think this was fixed by putting these methods inside the `de` mod. Thanks a ton @Kixunil and @tcharding for holding my hand. Net deletion!

![image](https://github.com/user-attachments/assets/a7dcb55a-7d4c-4e1c-828c-4093fabc61a9)

Kixunil commented on src/lib.rs:
```diff
@@ -136,15 +136,14 @@ impl<'a, NetVal: NetworkValidation, T> Uri<'a, NetVal, T> {
 
 impl<'a, T> Uri<'a, bitcoin::address::NetworkUnchecked, T> {
     /// Checks that the bitcoin network in the URI is `network`.
```
Actually, this comment was wrong as it suggests checking equality which would be wrong.

I'd prefer to update it before releasing next version.

Kixunil reviewed (commented):
ACK except either last commit, leaving comment update and release for later or doing the comment change within the PR.

Also if you decide to edit the PR you might as well reorder the commits to move the impl block first and update `bitcoin` sencond. I find it a tiny bit clearer to not change an API twice in a row.

Kixunil commented on src/de.rs:
```diff
@@ -228,6 +228,7 @@ enum UriErrorInner {
     TooShort,
     InvalidScheme,
     Address(AddressError),
+    AddressNetwork(bitcoin::address::Error),
```
Oh crap, I didn't realize reordering commits would require this. Anyway, let's not mess with it anymore.

DanGould replied:
I was gonna leave a comment but this is a minor historical flaw tbh

Kixunil replied:
My point with reordering was to avoid funny things like this. So we've avoided one and created another. :man_shrugging:
